### PR TITLE
Feature: Option to remove server side auth

### LINKED ISSUE #551: When refresh browser on prerender env(SSG), Authentication status is lost.
### Environment

------------------------------
- Operating System: Darwin
- Node Version:     v18.17.1
- Nuxt Version:     3.7.4
- CLI Version:      3.9.0
- Nitro Version:    2.6.3
- Package Manager:  npm@9.6.7
- Builder:          -
- User Config:      devtools, build, alias, ssr, runtimeConfig, nitro, vite, modules, auth, i18n, experimental, css, pinia, app
- Runtime Modules:  @pinia/nuxt@0.4.11, @pinia-plugin-persistedstate/nuxt@1.1.2, @nuxtjs/i18n@8.0.0-rc.5, @sidebase/nuxt-auth@0.6.0-rc.0
- Build Modules:    -
------------------------------
client : npm run generate > npx serve [output static directory] -l 3001
node server :  npm run build(port 3000) > node index.mjs

### Reproduction

1. Call signIn on Login Page.
``` javascript
        const { signIn, status } = useAuth();
	setUser(<resLoginType>returnData.value.rs);
	// await authLogin();
	const wrapBody = {
		username: loginId.value,
		password: passwd.value,
		
	};
	await signIn(wrapBody, { redirect: false });

	navigateTo('/main', { replace: true });
```
2. signin is authorize and then move main page. (Screenshots : Chrome Inspector Application > Cookie and check useAuth status value) 
스크린샷 2023-10-18 오후 6 02 46
스크린샷 2023-10-18 오후 6 09 34

3. After refreshing the browser, useAuth status value is lost and move the login.
스크린샷 2023-10-18 오후 6 18 48

### Describe the bug

1. Authentication passed on the login page.
2. Refreshed the main page.
3. Authentication status is lost and moved to the login page.

I guess this is probably a problem that occurs when pre-rendering is performed on the client-driven server side.
Is there a solution to this?

### Additional context

 "nuxt.config.ts" part the below.
``` typescript
import { resolve } from 'node:path';

// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
	devtools: { enabled: true },
	build: {
		transpile: ['vuetify', 'jsonwebtoken'],
	},
	alias: {
		'./runtimeConfig': './runtimeConfig.browser',
		cookie: resolve(__dirname, 'node_modules/cookie'),
	},

	ssr: true,
	runtimeConfig: {
		public: {
			baseApiUrl: process.env.BASE_API_URL,
			midAwsUrl: process.env.MIDDLEWARE_LAMBDA_URL,
			mode: process.env.ENV_MODE,
		},		
	},

	nitro: {
		preset: 'node-server', //'aws-lambda',
		storage: { '.data:auth': { driver: 'fs', base: './.data/auth' } },
		routeRules: {
			'/': { ssr: true, prerender: true },
		},		
	},
	vite: {
		envDir: process.cwd(),
		define: {
			'process.env.DEBUG': false,
			'window.global': {},
		},
		css: {
			preprocessorOptions: {
				scss: {
					additionalData: '@import "@/assets/main.scss"; ',
				},
			},
		},
	},
	modules: [
		'@pinia/nuxt',
		'@pinia-plugin-persistedstate/nuxt',
		'@nuxtjs/i18n',
		'@sidebase/nuxt-auth',
	],
	auth: {
		provider: {
			type: 'local',
			token: {
				signInResponseTokenPointer: '/token/accessToken',
				maxAgeInSeconds: 18000,
				// sameSiteAttribute: 'lax',
			},
		},
		baseURL: 'http://localhost:3000/api/auth', 

		globalAppMiddleware: {
			isEnabled: true,
		},
	},
	i18n: {
		strategy: 'no_prefix',
		locales: [
			{
				code: 'vi',
				name: 'Vietnamese',
				file: 'vi-VN.ts',
			},
			{
				code: 'en',
				name: 'English',
				file: 'en-US.ts',
			},
			{
				code: 'ko',
				name: 'Korean',
				file: 'ko-KR.ts',
			},
		],
		experimental: {
			jsTsFormatResource: true,
		},
		lazy: true,
		langDir: 'lang',
		defaultLocale: 'en',
	},
	experimental: {
		payloadExtraction: false,
	},
	css: ['vuetify/styles', '@/assets/dev.scss'],
	pinia: {
		autoImports: [
			// `defineStore`를 자동으로 임포트함.
			'defineStore', // import { defineStore } from 'pinia'
			['defineStore', 'definePiniaStore'], // import { defineStore as definePiniaStore } from 'pinia'
		],
	},
	app: {
		head: {
			title: 'D-Sales Store',
			// titleTemplate: 'input title template',
			link: [
				{ rel: 'stylesheet', href: '' },
				{ rel: 'preconnect', href: '' },
				{ rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' },
				// font - NotoSans 추가
				{
					rel: 'stylesheet',
					type: 'text/css',
					href: 'https://fonts.googleapis.com/css?family=Noto+Sans+KR:400,500,700',
				},
			],
		},
	},
});

```

### Logs

_No response_

### PR DESCRIPTION
**Pull request by KyleSmith0905**
~Closes~ Contributes to #551.

This adds an option `disableServerSideAuth` that when enabled, would make `useAuth` calls not pass down the user's session on the server.

For example: If a website has caching, we don't want to send the user a page that says they're logged in as someone else.

Checklist:
- [x] issue number linked above after pound (`#`)
    - replace "Closes " with "Contributes to" or other if this PR does not close the issue
- [x] manually checked my feature / checking not applicable
- [x] wrote tests / testing not applicable
- [x] attached screenshots / screenshot not applicable

### CODE REVIEW

**Comment by zoey-kaiser** [comment]
Hi @KyleSmith0905 👋 

Thanks for your PR. Could you please provide more context on when this is needed and what your current setup looks like, that make it required to have this in place? Thanks!

**Comment by KyleSmith0905** [comment]
## What is the use case?
I have an app with ISR/prerender on most routes. It's rendered on the server and the response is cached on the edge. With what we currently have (and after this update if `disableServerSideAuth` is not true), the cached response may contain a user's authentication session that may be entirely separate from the user making the request.

Here is a repo I created to share the problem: https://github.com/KyleSmith0905/mre--nuxt-app/tree/using-isr

## Why is it a global setting?
~It was easier to implement~ Many times if you're using ISR another type of caching, it's likely going to be on other pages as well I guess.

## Current implementation
I added a `disableServerSideAuth` option that prevents the server from getting the user session. It is by default `false`.
```ts
export default defineNuxtConfig({
  ...
  auth: {
    provider: {
      type: 'authjs'
    },
    disableServerSideAuth: true,
  }
})
```

**Comment by zoey-kaiser** [comment]
Hi @KyleSmith0905 thanks for the detailed response!

Ill properly test this tomorrow! Do you also want to do me a favor and add it to the docs? That way I can see if I still get this merged for you before the new year!

**Comment by KyleSmith0905** [comment]
Sounds good

I added to the JSDoc documentation on the Nuxt Config page to https://sidebase.io/nuxt-auth/configuration/nuxt-config
I couldn't find any other places that fit 100% comfortably, besides putting it in a tip page on the getting started page?

**Comment by brandon-gs** [comment]
Hello there @KyleSmith0905 

Can this work if I only have my root page using swr and cached with cloudflare, but other pages using SSR?

**Comment by KyleSmith0905** [comment]
It would make every page use client side auth, sorry. :/

I figured that if someone was using caching strategies, they'd have caching on many more places.

I'm not familiar with playing around with routeRules, but I can envision that I can put a property in there for disabling server side auth.

I agree that it should be done that using routeRules though. So I might work on the weekend to try to see if it would work.

**Comment by brandon-gs** [comment]
I would greatly appreciate it if you could do that! I want to implement this in a project that is already quite large and has too many routes, so I would like to progressively implement prerendering (SSG) or SWR on pages where I see it necessary and leave some using SSR if required. So using the disableServerSideAuth option from routeRules would be extremely helpful for me!

**Comment by KyleSmith0905** [comment]
- Added an `auth` object to `routeRules`. I kept in the global option I had before. The route rules takes priority over the module rules.

- Added a cached page in the playground project for all of the 3 providers to demonstrate route-based rules.

- Added a lot more documentation (a whole separate page).

If there are any issues, let me know. I changed a lot of lines of code. I couldn't find anything breaking.

**Comment by brandon-gs** [comment]
Thank you so much for your effort, @KyleSmith0905!

It appears that `test-playground-auth` is failing because, in `playground-authjs/nuxt.config.ts`, inside the auth configuration, the port in `baseURL` should be set to 3001.

**Comment by zoey-kaiser** [comment]
Thanks @KyleSmith0905 for continuing to push this into the new year!
I plan on reviewing + hopefully merging next Monday. Meanwhile, @brandon-gs do you think you could also do a review of the request you had? I think this PR is almost over the finish line!

**Comment by KyleSmith0905** [comment]
What's the next step in the review process?

**Comment by zoey-kaiser** [comment]
> What's the next step in the review process?

Hi @KyleSmith0905! I have passed this onto @BracketJohn to review and decide if this fits into our vision of NuxtAuth. Sadly he is pretty busy at the moment, resulting in some delays. 

I cannot give you an ETA at the moment. Is this PR super pressing for you? If it is, I could look into making a beta release with this PR for you and then we do a retrospective review (I would obviously not prefer this, but it is one of the options).

**Comment by KyleSmith0905** [comment]
That's fine, I can wait. I just wanted to make sure it wasn't entirely forgotten. All is good.

File: src/runtime/composables/local/useAuthState.ts [hunk 1]
```diff
@@ -44,11 +44,22 @@ export const useAuthState = (): UseAuthStateReturn => {
     rawToken
   }
 
+  onMounted(() => {
```
Comment (phoenix-ru) [inline src/runtime/composables/local/useAuthState.ts:47]: Also haven't validated this one yet. I think the PR needs a good functional review in general

File: src/runtime/types.ts [hunk 2]
```diff
@@ -292,6 +292,14 @@ export interface ModuleOptions {
    * Whether the module is enabled at all
    */
   isEnabled?: boolean;
+  /**
+   * Forces your server to send a "loading" status on all requests, prompting the client to fetch on the client. If your website has caching, this prevents the server from caching someone's authentication status.
+   *
+   * This effects the entire site, for route-specific rules, add `disableServerSideAuth` on `routeRules`.
```
Comment (phoenix-ru) [inline src/runtime/types.ts:298]:
```suggestion
   * This affects the entire site. For route-specific rules add `disableServerSideAuth` on `routeRules` instead:
      ```ts
      defineNuxtConfig({
        routeRules: {
          '/': { disableServerSideAuth: true }
        }
      })
      ```
```

**Review by phoenix-ru — CHANGES_REQUESTED** [review]

**Comment by KyleSmith0905** [comment]
I truly appreciate your feedback. Your feedback showed much greater foresight than I've considered when initially developing it.

Also, I believe I'm ready for another review when you're free.

I created live deployments if you would like to see how they function.
**Auth.js provider**: https://nuxtauth-authjs.vercel.app/with-caching
**Local provider**: https://nuxtauth-local.vercel.app/with-caching

Would be neat to have Vercel deployments integrated into the CI, though I don't know if there are any caveats (I can delete my projects if you want to use the project ids nuxtauth-authjs and nuxtauth-local).

**Comment by zoey-kaiser** [comment]
Hi @KyleSmith0905 👋 

Thanks for all your great work (and the demos) we really appreciate it! It makes reviewing so much easier ❤️ 
I will do a functional review, based on your demos, but as @phoenix-ru has already begun the code review process, I will leave the code review to him. He is currently on vacation and will be back next week.

He can then get around to a review around Thursday and we can then look to (_finally_) merge your PR 💪

**Review by zoey-kaiser — APPROVED** [review]
Functionally approved 😊

File: src/runtime/utils/kit.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  const options: Partial<RouteOptions> = {}
+
+  const matches = routeMatcher.matchAll(
+    withoutBase(withoutTrailingSlash(withoutQuery(path)), app.baseURL)
+  ).reverse()
+
+  for (const match of matches) {
+    options.disableServerSideAuth ??= match.auth.disableServerSideAuth
```
Comment (phoenix-ru) [inline src/runtime/utils/kit.ts:41]: `auth` may be undefined
```suggestion
    options.disableServerSideAuth ??= match.auth?.disableServerSideAuth
```

**Comment by KyleSmith0905** [comment]
Thanks, for the reviews. I've made edits accordingly.

I'm using a different branch for the demo URLs, not sure what the problem was with those, but I deployed again and fixed it.

**Review by phoenix-ru — APPROVED** [review]

**Comment by phoenix-ru** [comment]
@KyleSmith0905 Thank you for your PR! It was successfully merged and we will probably schedule it for the next patch release.

**Pull request merged**